### ISSUE TITLE
Hexagonal and other shapes for Halos

### ISSUE THREAD

**Issue by bubblobill** [RPTools/maptool#1616]
**Is your feature request related to a problem? Please describe.**
Halos are square. On maps with hexagonal grids this looks a little off.
![image](https://user-images.githubusercontent.com/45483160/79682771-85655e00-8257-11ea-8d3f-0f765e0d5a90.png)

**Describe the solution you'd like**
Have halos follow the grid system in use, or be definable in campaign properties with alternate shapes such as a circle.

**Comment by Merudo** [comment]
What shape should the halo be when the token is large/huge/humongous?

**Comment by Merudo** [comment]
Also to be considered: iso-grid halos are too large for rotated top-down tokens:

![image](https://user-images.githubusercontent.com/7371085/83933427-fdd3a000-a76d-11ea-9b79-661eeabc34a6.png)

**Comment by Phergus** [comment]
The halo's need to be based on footprint which should be the area encompassed by a group of contiguous cells.  The halo should be drawn on the perimeter of the footprint.  The footprint is not the same as the extent of the token image.  A user should be able to place a top down image that is irregular in shape and place it over a footprint with parts of the image hanging over the edges of the footprint.  When selecting a token the footprint should be highlight and not the rectangular bounds of the image.

Freesize tokens are a different beast.  The footprint should be the non-transparent area of the image with the halo around that.

Unfortunately the system put in for footprints years ago was based on a Cartesian coordinates and falls apart on hex grids when tokens are moved or rotated. For hex grids, the cells that make up the footprint need to be defined as vectors with direction and distance so that they can rotated around the pivot point.

Some example Hex footprints:

![Hex Grid Footprints](https://user-images.githubusercontent.com/34379254/83950703-04076200-a7ea-11ea-8f3d-86a497447454.png)

**Comment by Merudo** [comment]
Thanks for the useful grid picture, @Phergus!

I like the "halo around the footprint" idea. I assume halos on square grids should work like this too?

Regarding non snap-to-grid tokens, maybe the halo should be the same as for snap-to-grid tokens, except with an offset.

Regarding the picture, currently only 1 (Medium), 3 (Large), 7 (Huge), and 19 (Humongous) are implemented.

**Comment by Phergus** [comment]
> I like the "halo around the footprint" idea. I assume halos on square grids should work like this too?

Yes.  Halos should be based on footprint.  Essentially working exactly as they do today for the square footprint tokens.  What's missing is footprints for non-square entities such as vehicles which might be 1x2, 2x3, 2x4, etc..

Yes, the current hex footprints are just the simplistic implementation and if you try the 3 hex footprint you'll see it doesn't handle facing at all.

**Comment by Merudo** [comment]
> Yes. Halos should be based on footprint. Essentially working exactly as they do today for the square footprint tokens.

Square tokens don't always work this way:

![image](https://user-images.githubusercontent.com/7371085/83952075-cf9ca180-a7fb-11ea-99b4-b08609022ec1.png)

Here is a token with a 1 square footprint, but zoomed up in the layout manager. As you can see, the yellow aura is larger than 1 square and is the size of the token image.

**Comment by JamzTheMan** [comment]
The Grid class does have a "cell shape" and IIRC I used it to create GRID based lighting.  It shouldn't be that hard I would think to extend that to the halo and footprint code.

But we also have "foot print" code/logic in way to many places, duplicated if you will, so it could be a big refactor. Halo's at least should be easier? Never looked at them though...

**Comment by Phergus** [comment]
> Square tokens don't always work this way:

True but that's because currently halos are tied to image size/shape and not footprint.  Of course, if you have a token with a single cell footprint but have an image that is far larger then all bets are off the table.  For well-behaved portrait style tokens, the halo does match the footprint.

Related to that is how token states and bars can be drawn way off from actual token because the token image includes a tail or spear.

The point of footprints is to show what grid cells they occupy to facilitate game mechanics that depend upon knowing which grid cells an entity occupies or passes through.

**Comment by bubblobill** [comment]
Reraising the dead. Different idea, make a halo a halo. A filter effect. It will then follow the shape of the transparency.
You could produce it with 
https://www.tutorialspoint.com/javafx/javafx_effects.htm
multiple drop-shadows, bloom, &/or glow.
